### DISCUSSION TITLE
Hilfe mit Timer in timer app

### DISCUSSION THREAD

**Discussion by PhilipZi** [neuefische/questions#65] [Web] [answered]
Hallo zusammen, 

ich arbeite gerade an einer timer app und habe den timer fertig gecodet, allerdings weiss ich nicht wie ich diesen timer manuell anpassen kann, dass ich die Sekunden bzw. die Minuten erhöhen bzw, vermindern kann. Ich arbeite mit next.js und weiss leider nicht mehr weiter. 
Ich hoffe mir kann jemand helfen, bzw. eine gute Ressource zur Verfügung stellen. 

Liebe Grüsse

**Comment by alexclaes** [comment]
Hallo Philip,

hast du einen Link zu dem Code, den du bereits geschrieben hast? Kannst du erklären an welcher Stelle du nicht weiter kommst.

Im Grunde benötigst du zwei Dinge:
- Einen State, um die gewählte Dauer zu speichern (lokales useState oder globalen State mit zustand.js)
- Ein Eingabefeld, um den Wert zu setzen. Bei dir würde sich `input type="range"` anbieten https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range

Viele Grüße,
Alex

**Reply by PhilipZi** [reply]
Hey Alex, 

danke für deine Unterstützung, ich habe einen Code, der sehr lang ist und ich bin mir nicht sicher ob es möglich ist in diesem einen input einzubauen. Ich lade Ihn mal hier hoch. 
Ich glaube so wie ich den Code geschrieben habe kann ich an den Sekunden so oder so nichts verändern. Ich weiss leider wirklich nicht weiter.  Über jeden Hilfe bin ich sehr froh! Danke nochmal! `

export default function Timer() {
	const [thirtysec, setThirtysec] = useState(1);
	const [onemin, setOnemin] = useState(5);
	const [fivemin, setFivemin] = useState(20);
	const [seconds, setSecond] = useState(0);
	const [stage, setStage] = useState(0);
	const [ticking, setTicking] = useState(false);

	const switchStage = index => {
		setStage(index);
	};

	const getTickingTime = () => {
		const timeStage = {
			0: thirtysec,
			1: onemin,
			2: fivemin,
		};
		return timeStage[stage];
	};

	const updateMinute = () => {
		const updateStage = {
			0: setThirtysec,
			1: setOnemin,
			2: setFivemin,
		};
		return updateStage[stage];
	};

	const clockTicking = () => {
		const minutes = getTickingTime();
		const setMinutes = updateMinute();

		if (minutes === 0 && seconds === 0) {
			alert('Time up');
		} else if (seconds === 0) {
			setMinutes(minute => minute - 1);
			setSecond(59);
		} else {
			setSecond(second => second - 1);
		}
	};

	useEffect(() => {
		const timer = setInterval(() => {
			if (ticking) {
				clockTicking();
			}
		}, 1000);

		return () => {
			clearInterval(timer);
		};
	}, [seconds, thirtysec, onemin, fivemin, ticking]);

	return (
		<div>
			<Times
				stage={stage}
				switchStage={switchStage}
				getTickingTime={getTickingTime}
				seconds={seconds}
				ticking={ticking}
				setTicking={setTicking}
			/>
		</div>
	);
}
`
`export default function Times({stage, switchStage, getTickingTime, seconds, ticking, setTicking}) {
	const options = ['1min', '5min', '20min'];

	return (
		<>
			<div>
				{options.map((option, index) => {
					return (
						<h1 key={index} onClick={() => switchStage(index)}>
							{option}
						</h1>
					);
				})}
			</div>
			<div>
				<h1>
					{getTickingTime()}:{seconds.toString().padStart(2, '0')}
				</h1>
			</div>
			<button onClick={() => setTicking(ticking => !ticking)}>
				{ticking ? 'Stop' : 'Start'}
			</button>
		</>
	);
}
`

**Comment by PhilipZi** [comment]
ich bin mir auch nicht sicher ob es sich hier anbieten würde mit Zustand zu arbeiten? Bzw. möglich ist?

**Comment by PhilipZi** [comment]
das input type="range" ist auf jeden Fall genau das was ich umsetzten möchte! Danke für die Info, aber ob ich das in meinem vorhandenem Code implementieren kann, glaube ich leider nicht. Hast du da eine Idee?

**Comment by alexclaes** [comment, marked as answer]
Hi Philip,

es würde sich sehr anbieten den langen Code in einer IDE mit Syntax Highlighting lesen zu können ;-) Bei deinem Code ist auch etwas mit der Formatierung kaputt gegangen. Ein Link zu der Datei in deinem Branch auf GitHub wäre hilfreicher.

Hier mal ein Einsatz:

Wenn ich deinen Code richtig lese, setzt du in der `Timer` Komponente einen State mit den Zeiteinheiten:

```js
const [thirtysec, setThirtysec] = useState(1);
const [onemin, setOnemin] = useState(5);
const [fivemin, setFivemin] = useState(20);
const [seconds, setSecond] = useState(0);
```

Hier solltest du nicht die initialen Werte für den State hardcoden (1, 5, 20, 0), sondern der Komponente als Prop übergeben. Z.B. so

```js
export default function Timer({initialThirtysec, initialOnemin, initialFivemin, initialSeconds}) {
  // ...
}
```

Du nimmst den Werte aus dem Range Slider (z.B. eine gesamte Dauer in Sekunden). Diese rechnest du auf die von die genutzten  Einheiten um und übergibst sie an die `Timer` Komponente.

```jsx
<Timer initialThirtysec={1} initialOnemin={5} initialFivemin={20} initialSeconds={0} />
```

**Reply by PhilipZi** [reply]
Danke Alex, das ist genau das was ich meinte, danke! 
außerdem habe ich das Problem, dass in dieser Zeile des Codes die Sekunden wie ich glaube nicht anpassbar sind. Hättest du dafür auch eine lösung? 

```js
                         <div>
				<h1>
					{getTickingTime()}:{seconds.toString().padStart(2, '0')}
				</h1>
			</div>
```

ich habe hier nochmal einen Link zu meinem neu erstellten Pullrequest, so dass man sich den Code besser angucken kann [Github Pullrequest](https://github.com/PhilipZi/capstone-project/tree/edit-timer-card/src/components/timer) Und nochmal Tausend Dank!
